### PR TITLE
fix(zoom): making zoom in+out a usable feature

### PR DESCRIPTION
**Pull request by pongstr**
- feat: keyboard viewport addresses #115 
  - added: pan around when zoomed in+out
    - hold spacebar and move around
    - hold primary click button to move around
  - added: ui handlers to increase/decrease zoom and reset view to optimal settings
  - refactor: scaling should not require child components to scale too, no re-render while zooming in+out
  - refactor: `useLocalStorageState` to have default serializer+deserializer

https://github.com/user-attachments/assets/e84c239d-7e4c-4f5e-909f-f298a009495e

### CODE REVIEW

File: src/keyboard/Keyboard.tsx [hunk 1]
```diff
@@ -29,8 +29,7 @@ import { BehaviorBindingPicker } from "../behaviors/BehaviorBindingPicker";
 import { produce } from "immer";
 import { LockStateContext } from "../rpc/LockStateContext";
 import { LockState } from "@zmkfirmware/zmk-studio-ts-client/core";
-import { deserializeLayoutZoom, LayoutZoom } from "./PhysicalLayout";
-import { useLocalStorageState } from "../misc/useLocalStorageState";
```
Comment (Nicell) [inline src/keyboard/Keyboard.tsx:33]: `useLocalStorageState` was created for future local storage uses. I don't think we should remove it https://github.com/zmkfirmware/zmk-studio/pull/31#discussion_r1771781964

File: src/keyboard/KeyboardViewport.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
+import { FC, PropsWithChildren, useEffect, useRef } from "react";
+
+type KeyboardViewportType = PropsWithChildren<{
+  className?: string;
+}>;
+
+const KEYMAP_SCALE = "keymap:scale";
```
Comment (Nicell) [inline src/keyboard/KeyboardViewport.tsx]: Why are we renaming the key? This will make existing stored values invalid

File: src/keyboard/KeyboardViewport.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+const DEFAULT_SCALE = window.localStorage.getItem(KEYMAP_SCALE) ?? "1";
+
+export const KeyboardViewport: FC<KeyboardViewportType> = ({
```
Comment (Nicell) [inline src/keyboard/KeyboardViewport.tsx:12]: I disagree that this simplifies the scale localstorage in any meaningful way. This ref management is hard to follow, and without the deserialization we have to keep casting back and forth between numbers and strings.

File: src/keyboard/KeyboardViewport.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  };
+
+  useEffect(() => {
+    if (!targetRef.current) return;
+
+    const target = targetRef.current;
+    const offset = { x: 0, y: 0 };
+    let isPanningActive = false;
+
+    function panStart(e: KeyboardEvent) {
```
Comment (Nicell) [inline src/keyboard/KeyboardViewport.tsx]: So this only supports pan via spacebar? I think that's a good addition, but I think this should support click and drag, as most users will use that.

File: src/keyboard/KeyboardViewport.tsx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    function onInputChange(e: Event) {
+      const value = (e.currentTarget as HTMLInputElement).value;
+      target.style.setProperty("transform", `scale(${value})`);
+      localStorage.setItem(KEYMAP_SCALE, value);
+    }
+
+    input.addEventListener("change", onInputChange);
+    return () => {
+      input.removeEventListener("change", onInputChange);
```
Comment (Nicell) [inline src/keyboard/KeyboardViewport.tsx]: I think I noticed from your video that it only seems to change once you've let go. IIRC if you use react state + a on change handler in JSX it will update as you drag (which seems desirable). Overall not a fan of doing all this imperative work, which IMO should only be done when required (refs, event listeners, etc)

File: src/keyboard/KeyboardViewport.tsx (hunk 2 above)
Comment (pongstr) [reply src/keyboard/KeyboardViewport.tsx]: i should've marked it as a `TODO`, i didn't want to break anything that might be using the namespace. this can be renamed to what will be preferred.

File: src/keyboard/KeyboardViewport.tsx (hunk 4 above)
Comment (pongstr) [reply src/keyboard/KeyboardViewport.tsx]: we can add that as well.

edit: pointer down is added

File: src/keyboard/KeyboardViewport.tsx (hunk 5 above)
Comment (pongstr) [reply src/keyboard/KeyboardViewport.tsx]: revised the implementation

File: src/keyboard/KeyboardViewport.tsx (hunk 3 above)
Comment (pongstr) [reply src/keyboard/KeyboardViewport.tsx:12]: this has been addressed as well.

File: src/keyboard/Keyboard.tsx (hunk 1 above)
Comment (pongstr) [reply src/keyboard/Keyboard.tsx:33]: restored `useLocalStorageState`, i slightly modified the hook to have default serialize+deserializer, imho, these options should only be required for specific serialize+deserializer functions. lmk what you think ^^
